### PR TITLE
Use pre-commit package to do codestyle CI checks

### PR DESCRIPTION
**Pull request by Cadair**
The objective of this PR is to suggest a solution to the problem discussed at the coordination meeting of how can we make it easy for people to ensure that their code is conformant to the CIs enforcement of the style before they push.

This does two things:

1) Specify the code style checks using the [pre-commit](http://www.pre-commit.com/) package.
2) Run them in the `codestyle` tox env.

Using the pre-commit hooks locally as a git pre-commit is entirely optional, but doing so means you will *always* pass that CI build. If you don't want the git commit but want to check your code if conformant, you can run `pre-commit run --all-files`.

Also, I enabled a bunch of extra style checks, like trailing whitespace and blank lines at end of files etc. Fixing these is the large part of this diff.

### CODE REVIEW

**Comment by astrofrog** [comment]
Just to make sure I understand how this works, how does git know about this pre-commit hook?

File: tox.ini [hunk 1]
```diff
@@ -85,33 +85,10 @@ commands =
...
 skip_install = true
-description = check code style, e.g. with flake8
-deps = flake8
-commands = flake8 astropy --count --select=E101,W191,W291,W292,W293,W391,E111,E112,E113,E30,E502,E722,E901,E902,E999,F822,F823
+description = Run all style and file checks with pre-commit
+deps =
+    pre-commit
+commands =
+    pre-commit install-hooks
+    pre-commit run --all-files
```
Comment (bsipocz) [inline tox.ini]: couldn't this be the other way around, for pre-commit to run the tox command? 

E.g. looking into the tox config is more intuitive thing to do to look for the exact error codes than yet at another place.

Comment (astrofrog) [reply tox.ini]: I agree, is there any way to switch this around?

Comment (Cadair) [reply tox.ini]: That certainly doesn't work out of the box. Also pre-commit comes with a load of other checks plugins other than flake8, so it's a simple way to configure multiple checks.

**Comment by Cadair** [comment]
> Just to make sure I understand how this works, how does git know about this pre-commit hook?

If you want to teach git about it you can run
```
pre-commit install
```

you have to do this for each checkout you want to run the pre-commit for (git limitation).

**Comment by mhvk** [comment]
Not release-critical, so removing milestone

**Comment by nstarman** [comment]
@Cadair How can I help get this restarted / in ?

**Pull request closed without merging**